### PR TITLE
[debugger] Add features: print input and source

### PR DESCRIPTION
**Pull request by s123unny**

### CODE REVIEW

File: aeneas/src/ssa/SsaDebugger.v3 [hunk 1]
```diff
@@ -1,40 +1,56 @@
...
+	Next("next", "n", "Step program, jumping over calls"),
+	Nexti("nexti", "ni", "Step one instruction, jumping over calls"),
+	Finish("finish", "fin", "Execute until selected stack frame returns"),
+	Info("info", "info", "Show breakpoints or locals. Type \"help info\" for more help"),
+	Breakpoint("breakpoint", "b", "Making program stop at certain line: <fileName> <line>"),
+	DisableBreakpoint("disable", "disable", "Enable breakpoint by <index>"),
+	EnableBreakpoint("enable", "enable", "Disable breakpoint by <index>"),
+	Backtrace("backtrace", "bt", "Print a backtrace"),
+	Source("source", "source", "Print source with execution result along with a <line number>"),
+	Display("display", "display", "Print information during steping program. Type \"help display\" for more help"),
```
Comment (titzer) [inline aeneas/src/ssa/SsaDebugger.v3:19]: s/steping/stepping/

File: aeneas/src/ssa/SsaDebugger.v3 [hunk 2]
```diff
@@ -109,35 +139,48 @@ class SsaDebugger(prog: Program, interp: SsaInterpreter) {
...
 		var n = interp.step(interp.frame.ip);
 		if (interp.frame == null) {
 			running = false;
 			printExecuteResult();
 			return;
 		}
+		srcPrinter.putData(prev, targetSp);
 		interp.frame.ip = n;
-		while (!stopExecute(n, untilBreak, nextLine, jump, targetSp)) {
+		while (!stopExecute(n, untilBreak, nextLine, jump, fin, targetSp)) {
```
Comment (titzer) [inline aeneas/src/ssa/SsaDebugger.v3:163]: The four booleans make this a little difficult to figure out what combinations are valid. Does an enum make sense for the stopping condition?

File: aeneas/src/ssa/SsaDebugger.v3 [hunk 3]
```diff
@@ -147,23 +190,14 @@ class SsaDebugger(prog: Program, interp: SsaInterpreter) {
...
-					checkpoint.breakpoint = true;
-					bps[i] = SsaCPBreakpoint.new(checkpoint);
-					return true;
-				}
-			}
-		}
-		return false;
+		else return false;
 	}
+	// Print next instruction or the fianl result of the debugger program
```
Comment (titzer) [inline aeneas/src/ssa/SsaDebugger.v3:195]: typo

File: aeneas/src/ssa/SsaDebugger.v3 [hunk 4]
```diff
@@ -173,19 +207,176 @@ class SsaDebugger(prog: Program, interp: SsaInterpreter) {
...
+		var ctype = vstCompound.getDeclaredType();
+		var irClass = prog.ir.getIrClass(ctype);
+		var irMethod: IrMethod;
+		for (m in irClass.methods) {
+			if (m != null && m.source == vstMethod) {
+				irMethod = m;
+				break;
+			}
+		}
+		// Generate ssa if haven't
```
Comment (titzer) [inline aeneas/src/ssa/SsaDebugger.v3:327]: Should this handle the case where the method isn't found?

File: aeneas/src/ssa/SsaDebugger.v3 [hunk 5]
```diff
@@ -173,19 +207,176 @@ class SsaDebugger(prog: Program, interp: SsaInterpreter) {
...
+			var ta = if(irMethod.typeArgs != null, Arrays.prepend(receiver, irMethod.typeArgs.types), [receiver]);
+			var spec = prog.ir.fromIrMemberAndTypes(ta, irMethod);
+			irMethod.ssa = genSsa(spec, 0);
+		}
+		// Iterate ssa to find SsaCheckpoint pointing to line
+		var blocks = irMethod.ssa.bfBlocks(null);
+		for (b in blocks) b.clearMark();
+		for (b in blocks) {
+			for (i = b.next; i != null && i != b; i = i.next) {
+				if (SsaCheckpoint.?(i) && SsaCheckpoint.!(i).source.line == line) {
```
Comment (titzer) [inline aeneas/src/ssa/SsaDebugger.v3:339]: You could do a `match(i) { x: SsaCheckpoint => { ... } }` here.

File: aeneas/src/ssa/SsaDebugger.v3 [hunk 6]
```diff
@@ -173,19 +207,176 @@ class SsaDebugger(prog: Program, interp: SsaInterpreter) {
...
+		// Generate ssa if haven't
+		if (irMethod.ssa == null) {
+			var receiver = vstCompound.declType;
+			var ta = if(irMethod.typeArgs != null, Arrays.prepend(receiver, irMethod.typeArgs.types), [receiver]);
+			var spec = prog.ir.fromIrMemberAndTypes(ta, irMethod);
+			irMethod.ssa = genSsa(spec, 0);
+		}
+		// Iterate ssa to find SsaCheckpoint pointing to line
+		var blocks = irMethod.ssa.bfBlocks(null);
+		for (b in blocks) b.clearMark();
```
Comment (titzer) [inline aeneas/src/ssa/SsaDebugger.v3:336]: Is `clearMark()` necessary here?

File: aeneas/src/ssa/SsaDebugger.v3 [hunk 7]
```diff
@@ -235,38 +426,150 @@ class DebuggerParser {
...
+	}
+	def putVal(instr: SsaInstr) {
+		var vtype = instr.getType();
+		if (VoidType.?(vtype) || instr.valueNum < 0) return;
+		var source = SsaApplyOp.!(instr).source;
+		var point = source.column - 1;
+		var pad = if(source.startColumn != -1, source.startColumn-1, point);
+		extractTmpConstraint(pad);
+		tmpBuf.pad(' ', pad);
+		tmpBuf.pad('~', point);
```
Comment (titzer) [inline aeneas/src/ssa/SsaDebugger.v3:543]: I was thinking of centering the value with a format of `|----(val)----|` or `^^^^ val ^^^^`. Wdyt?

File: aeneas/src/ssa/VstSsaGen.v3 [hunk 8]
```diff
@@ -353,17 +356,17 @@ class VstSsaGen extends VstVisitor<VstSsaEnv, SsaInstr> {
 		loopEnv.loopStart = loopUpdate;
 		loopEnv.loopEnd = loopEnd;
 		// add the loop control expression
-		var cond = loopEnv.at(null).opIntLt(loopEnv[stmt.loopVar], limitExpr);
+		var cond = loopEnv.opIntLt(loopEnv[stmt.loopVar], limitExpr);
```
Comment (titzer) [inline aeneas/src/ssa/VstSsaGen.v3:359]: Can you factor out the `env.at(null)` changes to a separate PR? (Just to keep this PR smaller).

File: aeneas/src/ssa/SsaDebugger.v3 (hunk 7 above)
Comment (s123unny) [reply aeneas/src/ssa/SsaDebugger.v3:554]: I would like to use this one: `|----(val)----|`

File: aeneas/src/core/Value.v3 [hunk 9]
```diff
@@ -35,11 +35,16 @@ class Box<T>(val: T) extends Val {
 }
 // The source position in a method, such as for an exception stacktrace.
 class Source(mdecl: VstMethod, line: int, column: int, tail: Source) {
+	var startColumn = -1, endColumn = -1;
```
Comment (titzer) [inline aeneas/src/core/Value.v3:38]: Can you add these fields to the constructor and thus make them immutable?

**Review by titzer — APPROVED** [review]
lgtm with minor change

**Pull request merged**